network/ocsp.go

```diff
@@ -0,0 +1,273 @@
+/*
+Copyright 2020, Cossack Labs Limited
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package network
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/crypto/ocsp"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+)
+
+// --tls_ocsp_required=<denyUnknown|allowUnknown|all>
+const (
+	// Deny certificates now known by OCSP server(s)
+	ocspRequiredYesStr = "denyUnknown"
+	// Allow certificates now known by OCSP server(s)
+	ocspRequiredNoStr = "allowUnknown"
+	// Effect of denyUnknown + all available OCSP servers (the one from config
+	// and those listed in certificate) should respond, otherwise deny the certificate
+	ocspRequiredAllStr = "all"
+)
+
+const (
+	ocspRequiredYes int = iota
+	ocspRequiredNo
+	ocspRequiredAll
+)
+
+// --tls_ocsp_from_cert=<use|trust|prefer|ignore>
+const (
+	// Use OCSP servers listed in certificate (if any), try them after the one
+	// configured from CLI/config
+	ocspFromCertUseStr = "use"
+	// Query servers listed in certificate and don't perform further requests
+	// if one respons with "ok, valid"
+	ocspFromCertTrustStr = "trust"
+	// Query servers listed in certificate before the one from config
+	ocspFromCertPreferStr = "prefer"
+	// Ignore OCSP servers listed in certificates
+	ocspFromCertIgnoreStr = "ignore"
+)
+
+const (
+	ocspFromCertUse int = iota
+	ocspFromCertTrust
+	ocspFromCertPrefer
+	ocspFromCertIgnore
+)
+
+// OCSPConfig contains configuration related to certificate validation using OCSP
+type OCSPConfig struct {
+	url      string
+	required int // ocspRequired*
+	fromCert int // ocspFromCert*
+}
+
+// NewOCSPConfig creates new OCSPConfig
+func NewOCSPConfig(uri, required, fromCert string) (*OCSPConfig, error) {
+	if uri != "" {
+		_, err := url.Parse(uri)
+		if err != nil {
+			return nil, err
+		}
+
+		// TODO: Do some request to `uri`, log warn if failed
+	}
+
+	var requiredVal int
+	switch required {
+	case ocspRequiredYesStr:
+		requiredVal = ocspRequiredYes
+	case ocspRequiredNoStr:
+		requiredVal = ocspRequiredNo
+	case ocspRequiredAllStr:
+		requiredVal = ocspRequiredAll
+	default:
+		return nil, errors.New("Invalid `ocsp_required` value '" + required + "', should be one of 'yes', 'no', 'all'")
+	}
+
+	var fromCertVal int
+	switch fromCert {
```

Comment:
and same for other switches. here we should only check that it's valid value by `requiredVal, ok := requiredValueMap[requiredVal]` and check `ok` variable.